kategory-effects/src/main/kotlin/kategory/effects/instances/JobW.kt

```diff
@@ -0,0 +1,89 @@
+package kategory
+
+import kotlinx.coroutines.experimental.CoroutineStart
+import kotlinx.coroutines.experimental.Job
+import kotlinx.coroutines.experimental.launch
+import java.util.concurrent.atomic.AtomicReference
+import kotlin.coroutines.experimental.CoroutineContext
+
+@higherkind data class JobW<out A>(val thunk: ((Either<Throwable, A>) -> Unit) -> Job) : HK<JobWHK, A> {
+
+    fun <B> map(f: (A) -> B): JobW<B> =
+            JobW { ff: (Either<Throwable, B>) -> Unit ->
+                thunk { either: Either<Throwable, A> ->
+                    ff(either.map(f))
+                }
+            }
+
+    fun <B> flatMap(f: (A) -> JobW<B>): JobW<B> =
+            JobW { ff: (Either<Throwable, B>) -> Unit ->
+                val state = AtomicReference<A?>()
+                thunk { either: Either<Throwable, A> ->
+                    either.fold({}, { state.set(it) })
+                }.apply {
+                    this.invokeOnCompletion { t: Throwable? ->
+                        val a: A? = state.get()
+                        if (t == null && a != null) {
+                            f(a).thunk { either: Either<Throwable, B> ->
+                                ff(either)
+                            }
+                        }
+                    }
```

Comment:
There were two possible approaches to this impl. The first one requires a runBlocking { } block, the other uses invokeOnCompletion and a simple AtomicReference. Code-wise it seemed like the fastest approach.

I'd like @elizarov to weight in if we could get his advice.